persistence/trees/trees.go

```diff
@@ -251,6 +272,66 @@ func (t *treeStore) getStateHash() string {
 	return hexHash
 }
 
+////////////////////////////////
+// AtomicStore Implementation //
+////////////////////////////////
+
+// Savepoint generates a new savepoint for the tree store and saves it internally.
+func (t *treeStore) Savepoint() error {
+	w, err := t.save()
+	if err != nil {
+		return err
+	}
+	t.Prev = w
+	return nil
+}
+
+// Rollback intentionally can't return an error because at this point we're out of tricks
+// to recover from problems.
+func (t *treeStore) Rollback() {
+	if t.Prev != nil {
+		t.merkleTrees = t.Prev.MerkleTrees
+		t.rootTree = t.Prev.RootTree
+		return
+	}
+	t.logger.Fatal().Msgf("rollback called without valid savepoint - this is a bug - treeStore shutting down: %+v", t)
+}
+
+// save commits any pending changes to the trees and serializes the treeStore to a WorldState object
+// with new readers and writers.
+// OPTIMIZE: Consider saving only the root hash of each tree and the tree directory here and then
+// load the trees up in Rollback instead of setting them up here.
+func (t *treeStore) save() (*Worldstate, error) {
+	if err := t.Commit(); err != nil {
+		return nil, err
+	}
+
+	w := &Worldstate{
+		TreeStoreDir: t.treeStoreDir,
+		MerkleTrees:  map[string]*stateTree{},
+	}
+
+	for treeName := range t.merkleTrees {
+		root, nodeStore := t.GetTree(treeName)
+		tree := smt.ImportSparseMerkleTree(nodeStore, smtTreeHasher, root)
+		w.MerkleTrees[treeName] = &stateTree{
```

Comment:
Given that the directory where they are stored is the same (with the same name), how do we avoid overrides?

Reply:
I'm not sure what you mean by overrides. BadgerDB will only allow one connection at a time to a file, and will error if it detects a second attempted connection, is that what you mean?

Reply:
Yes, that's what I meant. I guess it's a non-issue?

Reply:
Unless I misunderstand the problem, I believe it's a non-issue. Each `nodeStore` handles the badgerDB reference, which can't be overridden because badgerDB will panic the caller that attempts a second connection.